New loader: generate dataframe

mariehbourget opened the pull request:
Fixes #519 
This PR is part of the loader refactoring project to include microscopy data (#304) and to migrate to HDF5 loader (#474).
It may also include changes to support EEG modality (#525).

It aims to generate a dataframe by indexing filenames and metadata in a bids path.

Changes have been tested with `anat` and `microscopy` datasets.

The changes are:
1 new function `create_bids_dataframe` in `ivadomed/loader/utils.py`.
1 new requirement: `pybids>=0.12.4`
3 new files:
`ivadomed/config/config_bids.json` : custom bids_config file needed for dealing with additional microscopy entities in `pybids`.
`ivadomed/config/config_new_loader.json` : ivadomed config file with 2 new `loader parameters`:  `bids_config` and `extensions`.
`dev/df_new_loader.py` : Script used to test the new `create_bids_dataframe` function.

I left a lot of comments in the code to explain the different steps and choices made.

I was able to keep the validation (by the BIDS-validator) by forcing the indexing of files in `sub-*` folders at the root of the bids path and of files specific to microscopy (`samples.tsv`, `samples.json`).

To discuss:
- Do we want to index derivatives? (depends on where and when we want to check for availability and do the dataset split)
- Currently, the files are indexed and filtered by `extensions` and `target_suffix`. We could also filter by `data_type` (anat, microscopy, etc) and/or modality suffix (ex: T1w, SEM, etc) in case multiple data type are available in a dataset.

TODO:
- Include parsing of `sessions.tsv` and `scans.tsv` files (not done at the moment, only `participants.tsv` and `samples.tsv`)
- Read appropriate metadata from image file and add to dataframe. Is it pertinent at this stage? Could also be done after the split when loading images to numpy arrays.
- Evaluate changes to make to include EEG.

mariehbourget committed:
requirements: Add pybids>=0.12.4

mariehbourget committed:
config: Add custom config_bids.json file

mariehbourget committed:
config: Add config_new_loader.json

mariehbourget committed:
loader/utils.py: Add create_bids_dataframe function

mariehbourget committed:
dev: Add df_new_loader.py script to test dataframe function

mariehbourget committed:
Fix typo

olix86 commented on ivadomed/config/config_new_loader.json:
```diff
@@ -0,0 +1,88 @@
+        "safety_factor": [1.0, 1.0, 1.0]
+    },
+    "loader_parameters": {
+        "bids_path": "/home/mhbourget/ivadomed/data_example_microscopy_sem",
```
same

mariehbourget replied:
Fixed in e5eb063

olix86 commented on ivadomed/config/config_new_loader.json:
```diff
@@ -0,0 +1,88 @@
+{
+    "command": "train",
+    "gpu": 0,
+    "log_directory": "/home/mhbourget/ivadomed/log_directory/new_loader",
```
I'd use a relative path such as
``` "log_directory": "log_directory/new_loader",```

that way it's easier for others to use it

mariehbourget replied:
Fixed in e5eb063

olix86 commented on ivadomed/loader/utils.py:
```diff
@@ -501,3 +503,81 @@ def reorient_image(arr, slice_axis, nib_ref, nib_ref_canonical):
+    df = pd.merge(df, df_participants, on='participant_id', suffixes=("_x", None), how='left')
+
+    # Add metadata from samples.tsv, if present
+    fname_samples = bids_path + "/samples.tsv"
```
Maybe consider [os.join](https://www.geeksforgeeks.org/python-os-path-join-method/)

mariehbourget replied:
Fixed in 7c10596

andreanne-lemay commented on ivadomed/loader/utils.py:
```diff
@@ -501,3 +503,81 @@ def reorient_image(arr, slice_axis, nib_ref, nib_ref_canonical):
+
+    Args:
+        loader_params (dict): Loader parameters, see :doc:`configuration_file` for more details.
+        derivatives (bool): If True, derivatives are indexed.
```
This is great :100: ! It makes it flexible for the dataset where we don't have derivatives. Concerning your question, I think we should index the derivatives if this option is set to True, it will be easier for the loading if everything is in the df. This function should also verify availability (in my opinion), so if we want to index derivatives, only the subjects with the right target(s) and contrast(s) should be in the final df.

mariehbourget replied:
Ok thanks, I also think it's better to check availability before the split, I'll look into that.

mariehbourget replied:
Filtering based on derivatives availability is done in b3c9d34. We now have to determine what is the behavior if no matching derivatives are found (currently raise an error and exit).

andreanne-lemay commented on ivadomed/loader/utils.py:
```diff
@@ -501,3 +503,81 @@ def reorient_image(arr, slice_axis, nib_ref, nib_ref_canonical):
+    # Initialize BIDSLayoutIndexer and BIDSLayout
+    # validate=True by default for both indexer and layout, BIDS-validator is not skipped
+    # Force index for subfolders starting with string "sub", and samples tsv and json files (for microscopy)
+    force_index = [d for d in os.listdir(bids_path) if (os.path.isdir(os.path.join(bids_path, d))) and (d.startswith("sub"))]
```
I understand why the indexing of 'samples.tsv' and 'samples.json' are forced, but why do we have to force the indexing of folders starting by "sub" won't they already be indexed by pybids?

mariehbourget replied:
This is because I kept the BIDS-validator "on", so files that are not BIDS-compliant are ignored. For microscopy, that means that sub folders at the root of the dataset are ignored because png and ometiff files are not compliant yet. So forcing "sub" is overkill for anat, but necessary for microscopy. When microscopy will be included in pybids and BIDS-validator, we will be able to remove the force_index.
This is not full proof because if non-compliant files are in the sub folders, they will be indexed to. I'll check if I can find a better condition.

mariehbourget replied:
I updated the condition in 12a5b50. The indexing is now forced for sub-folders only if they include microscopy files based on extensions.

andreanne-lemay commented on ivadomed/loader/utils.py:
```diff
@@ -501,3 +503,81 @@ def reorient_image(arr, slice_axis, nib_ref, nib_ref_canonical):
+    force_index = [d for d in os.listdir(bids_path) if (os.path.isdir(os.path.join(bids_path, d))) and (d.startswith("sub"))]
+    force_index.extend(['samples.tsv', 'samples.json'])
+    indexer = pybids.BIDSLayoutIndexer(force_index=force_index)
+    layout = pybids.BIDSLayout(bids_path, config=bids_config, indexer=indexer, derivatives=derivatives)
```
I see we have a fork of pybids, do we plan on overwriting some of their code? If so, we might want to handle the error when .json files exist, but are empty (https://github.com/ivadomed/ivadomed/issues/532). Currently, we still have an error in this case.

mariehbourget replied:
I was planning on making PR from the fork before I found the custom config file. For now, it seems easier to wait until BEP microscopy is merged before submitting changes to pybids. But the json file error has nothing to do with microscopy so yes, we could propose changes.

andreanne-lemay commented on ivadomed/loader/utils.py:
```diff
@@ -501,3 +503,81 @@ def reorient_image(arr, slice_axis, nib_ref, nib_ref_canonical):
+
+    # TODO: read appropriate metadata from image files and add to dataframe
+
+    return df
```
I tested it on MRI data and it worked very well! I love the final dataframe format!

mariehbourget replied:
Thanks!

andreanne-lemay commented on ivadomed/loader/utils.py:
```diff
@@ -501,3 +503,81 @@ def reorient_image(arr, slice_axis, nib_ref, nib_ref_canonical):
+
+    # Get bids_path, bids_config, target_suffix and extensions from loader parameters
+    bids_path = loader_params['bids_path']
+    bids_config = loader_params['bids_config']
```
We should make this parameter optional in the config file (we could simply put None by default) and if it is None, we take the default file present in ivadomed/ivadomed/config, because in most cases we don't need a custom bids file.

mariehbourget replied:
Fixed in 4e5eaf7
Let me know if this is what you had in mind.

andreanne-lemay reviewed (commented):
:1st_place_medal:

mariehbourget committed:
config: use relative paths

mariehbourget committed:
Define path with os.join

mariehbourget committed:
Remove unneccesary print

mariehbourget committed:
bids_config default to None if not present in config file

mariehbourget committed:
Change file filtering to work with relative paths and avoid df concat

mariehbourget committed:
Merge branch 'master' into mhb/519-df-new-loader

jcohenadad commented on ivadomed/loader/utils.py:
```diff
@@ -505,3 +507,78 @@ def reorient_image(arr, slice_axis, nib_ref, nib_ref_canonical):
     trans_orient = nib.orientations.ornt_transform(ras_orientation, ref_orientation)
     # apply transformation
     return nib.orientations.apply_orientation(arr_ras, trans_orient)
+
```
PEP8 asks for 2 lines between function defs

mariehbourget replied:
Fixed in 85537da

mariehbourget committed:
Read participants.tsv only if present

mariehbourget committed:
Correction PEP8

mariehbourget committed:
Minor corrections script dev

mariehbourget committed:
Config file: add "data_types" to loader params

mariehbourget committed:
Add filtering of dataframe by "data_types"

mariehbourget committed:
Merge branch 'master' into mhb/519-df-new-loader

mariehbourget committed:
Changes to parsing of participants.tsv and samples.tsv

mariehbourget committed:
Add parsing of sessions.tsv and scans.tsv

mariehbourget committed:
Add filtering of dataframe based on derivatives availability

mariehbourget committed:
Add filtering based on contrasts

mariehbourget committed:
Remove filtering based on data_type

mariehbourget commented:
Update in implementation

DONE:
-	Added filtering based on modality suffix (contrast)
-	Added filtering based on available derivatives
-	Added parsing or `sessions.tsv` and` scans.tsv` files

TODO:
-	Find a better condition for force index of microscopy files
-	Determine behavior if no matching derivatives are found (currently raise error and exit)
-	Load metadata from files (I'm not sure if it is the best place, should we do it in the loader specific to each file format instead?)
-	Create a test dataset including : `microscopy` (png + ometiff), `anat` (2D + 3D nifty), `dwi` (4D nifty), `eeg` (files TBD)
-	Adapt behavior of `scans.tsv` parsing for `eeg` and `dwi`.
-	Write a test for the function

mariehbourget committed:
Merge branch 'master' into mhb/519-df-new-loader

jcohenadad commented on ivadomed/loader/utils.py:
```diff
@@ -520,3 +522,116 @@ def reorient_image(arr, slice_axis, nib_ref, nib_ref_canonical):
+
+    # If indexing of derivatives is true
+    # Get list of subject files, list of subject prefix filenames without extensions, and list of available derivatives
+    if derivatives:
```
as discussed in meeting, you could inform with logger.debug if one derivative for a file is missing

mariehbourget replied:
Done in 7c3e8a2: warnings for missing one derivative for a subject, and missing all derivatives for a subject.

mariehbourget committed:
Merge branch 'master' into mhb/519-df-new-loader

mariehbourget committed:
Update force_index condition for microscopy files

mariehbourget committed:
Add warning when no derivatives exist for a subject file

mariehbourget committed:
Add warning when a target_suffix is missing for a subject file

mariehbourget commented:
This PR is ready for review.

As discussed:
- force index is now only applied to microscopy files
- if debugging is true: there are warnings for missing one derivative or all derivatives for a subject.

charleygros commented on ivadomed/config/config_new_loader.json:
```diff
@@ -0,0 +1,88 @@
+{
```
I am guessing we will use this file during the prototyping / development of the new loader before removing it when fully integrated. Is it the idea?

charleygros commented on ivadomed/config/config_new_loader.json:
```diff
@@ -0,0 +1,88 @@
+    },
+    "loader_parameters": {
+        "bids_path": "data_example_microscopy_sem",
+        "bids_config": "ivadomed/config/config_bids.json",
```
Does it mean this config_bids could potentially change across projects?

mariehbourget replied:
Not really. This custom config is a modification to the default config of `pybids` to include microscopy data, since microscopy won't be merged with pybids for a while. It includes everything that is already in the default pybids config file + relevant entities for microscopy. It could be updated if microscopy BEP changes, or if we want to include other "non-bids" modality.
For MRI and any other "official" bids modality, the custom pybids config file is not necessary and should not be used.

charleygros commented on dev/df_new_loader.py:
```diff
@@ -0,0 +1,39 @@
+#!/usr/bin/env python
```
Very well documented and useful. I am wondering if we should not move these lines to a proper testing function!

mariehbourget replied:
Yes absolutely! The purpose of this script was to test during development. I'll open another PR to include a proper testing function.

charleygros reviewed (approved):
Really well done @mariehbourget ! That was not a easy task to integrate this new hierarchy!
A small comment would be: to add some documentation about this new Galaxy in ReadTheDocs (e.g., re the new params in config file).

charleygros reviewed (commented):
I appreciated how well your code was commented / documented!
Can I help with something? eg run some tests? any idea?

mariehbourget committed:
Merge branch 'master' into mhb/519-df-new-loader

mariehbourget committed:
default config: Add "extensions" key

mariehbourget committed:
Correction to filtering when no extensions are provided

mariehbourget committed:
Doc: Add bids_config and extensions keys in configuration_file.rst

mariehbourget commented:
> Really well done @mariehbourget ! That was not a easy task to integrate this new hierarchy!
> A small comment would be: to add some documentation about this new Galaxy in ReadTheDocs (e.g., re the new params in config file).

Thank you @charleygros for your feedback. I added the new params in the config file. I also added the key "extensions" to the default config and modify how the filtering is done with the extensions in the function. That way, "extensions" is not mandatory for MRI (as it was before), and is necessary only if a specific dataset contains more than one possible file extensions (e.g. microscopy with png and ometiff). Could you review the changes? Thanks!

I was wondering if it is OK to merge the PR after that without the test, because the coverage will decrease, what do you think?
(I'm starting to work on the test dataset and test now, it will be in another PR)

charleygros commented on ivadomed/loader/utils.py:
```diff
@@ -570,12 +570,14 @@ def create_bids_dataframe(loader_params, derivatives):
+    # and with derivative files of chosen target_suffix from loader parameters
+    df = (df[(~df['path'].str.contains('derivatives') & df['suffix'].str.contains('|'.join(contrast_lst)) &
+          df['extension'].str.contains('|'.join(extensions))) |
+          (df['filename'].str.contains('|'.join(target_suffix)))])
```
:100:

charleygros commented:
> Thank you @charleygros for your feedback. I added the new params in the config file. I also added the key "extensions" to the default config and modify how the filtering is done with the extensions in the function. That way, "extensions" is not mandatory for MRI (as it was before), and is necessary only if a specific dataset contains more than one possible file extensions (e.g. microscopy with png and ometiff). Could you review the changes? Thanks!

I had a look at it: looks great! Many thanks for documenting this as well!

> I was wondering if it is OK to merge the PR after that without the test, because the coverage will decrease, what do you think?
> (I'm starting to work on the test dataset and test now, it will be in another PR)

Oh yeah, sure, makes complete sense! Let's first merge this PR, then introduce the testing dataset to properly test it. Cheers!

mariehbourget committed:
Merge branch 'master' into mhb/519-df-new-loader
